More descriptive error message for `uv run pythonx.xx`

JWLee89 opened the pull request:
## Related Issue 

https://github.com/astral-sh/uv/issues/11796

## Summary

1. Add meaningful message when user invokes uv with `run python<valid-python-version>`. 

 ✅ "python", "python3", "python3.9", "python4", "python3.10", "python3.13.3", "python39" (yes, python39 might still be released in 100 years 😆 )
 ❌ "python3abc", "python3.12b3", "", "python-foo"
 
 2. Contextual information added depending on whether the command was run inside of a `uv project` or not. See: https://github.com/astral-sh/uv/pull/12201/files#diff-423c3a82aab6dba4068dd7675391fc9328a66682131dbf2043e7847ab4ba8661R1092
 
 ## Automated Tests

1. Added unit test for new functions created for parsing and validating python executable entered by user
2. TODO: Once specs are finalized through discussions in PR, will create integration tests if necessary
 
 ## Need to Discuss

1. Invoking python via its patch version - E.g. `python3.11.9` is invalid. However, uv supports `uv run -p 3.11.9 python`, which is why when we do the following: 

```python
uv run python3.11.9
```

We will get the following error:

```shell
error: Failed to spawn: `python3.11.9`
  Caused by: `python3.11.9` not available in the environment, which uses python `3.11.9`. Did you mean to search for a Python 3.11.9 environment with `uv run -p 3.11.9 python`?
```

Is this okay? If not, we need to define the specs in the PR here.

2. Supported python executable specs are defined as: 
- is stable version
- is not `post` version

See:https://github.com/astral-sh/uv/pull/12201/files#diff-423c3a82aab6dba4068dd7675391fc9328a66682131dbf2043e7847ab4ba8661R1572-R1574

Would this be okay?
We can change the specs if needed. Lets discuss.

## Manual Test

### Case 1: run outside of project

1. Case where no project is created. Look for a python executable version that is not available in my local machine (in my case, it was `python3.9`)
4. Run `cargo run -- run python3.9.3`

```Shell
error: Failed to spawn: `python3.9.3`
  Caused by: `python3.9.3` not available in the environment, which uses python `3.11.9`. Did you mean to search for a Python 3.9.3 environment with `uv run -p 3.9.3 python`?
```

5.  Run `cargo run -- run python3.9`

```shell
error: Failed to spawn: `python3.9`
  Caused by: `python3.9` not available in the environment, which uses python `3.11.9`. Did you mean to search for a Python 3.9 environment with `uv run -p 3.9 python`?
```

6. Run `cargo run -- run python3.11`

```Shell
Running `target/debug/uv run python3.11`
Python 3.11.9 (main, Apr  2 2024, 08:25:04) [Clang 15.0.0 (clang-1500.3.9.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
```

7. Run `cargo run -- run python3.11.9`

```Shell
 Running `target/debug/uv run python3.11.9`
error: Failed to spawn: `python3.11.9`
  Caused by: `python3.11.9` not available in the environment, which uses python `3.11.9`. Did you mean to search for a Python 3.11.9 environment with `uv run -p 3.11.9 python`?
```

**IMPORTANT**: Is this acceptable? Invoking python by its patch version is not supported, so this is the expected behavior, but we need to discuss whether this behavior is acceptable.

### Case 2: run inside of project

Create project using `uv init example-app`

1. `uv run python3.13` 

```shell
error: Failed to spawn: `python3.13`
  Caused by: `python3.13` not available in the project environment, which uses python `3.12.5`. Did you mean to change the environment to Python 3.13 with `uv run -p 3.13 python`?
```

2. `uv run python3` 

```shell
uv run python3
Python 3.12.5 (main, Aug  6 2024, 19:08:49) [Clang 15.0.0 (clang-1500.3.9.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
```

JWLee89 committed:
feat: add function for checking if command is python executable

JWLee89 committed:
feat: Add meaningful message when attempting to invoke python using run

JWLee89 committed:
fix: enable message to show for patch versions

JWLee89 committed:
feat: add support for different message depending on if project is found

JWLee89 committed:
fix: run lint

JWLee89 committed:
fix: change message for project

JWLee89 committed:
docs: update docstring

zanieb commented:
Hey @JWLee89, thanks for the contribution! This looks like a nice improvement. We can provide some feedback to unblock adding tests.

zanieb commented on crates/uv/src/commands/project/run.rs:
```diff
@@ -1075,6 +1078,39 @@ hint: If you are running a script with `{}` in the shebang, you may need to incl
+            if err.kind() == std::io::ErrorKind::NotFound && is_python_executable(&executable) {
+                // Get version from python command string
+                // e.g. python3.12 -> "3.12" or "" if python version not specified. 
+                let version_part = executable.strip_prefix("python").unwrap_or("");
```
Should `is_python_executable` be `python_executable_version -> Some(str)` instead? Then you can avoid the `unwrap_or`?

JWLee89 replied:
Applied: https://github.com/astral-sh/uv/pull/12201/commits/0cb2d5e35dddaabf6eef93410643bbc58e8098fd

zanieb commented:
I think the semantics make sense here. I wonder if we can reuse anything from the `uv-python` crate, but nothing obvious comes to mind.

jtfmumm commented on crates/uv/src/commands/project/run.rs:
```diff
@@ -1075,6 +1078,39 @@ hint: If you are running a script with `{}` in the shebang, you may need to incl
+            if err.kind() == std::io::ErrorKind::NotFound && is_python_executable(&executable) {
+                // Get version from python command string
+                // e.g. python3.12 -> "3.12" or "" if python version not specified. 
+                let version_part = executable.strip_prefix("python").unwrap_or("");
```
Maybe `specified_version` instead of `version_part`?

JWLee89 replied:
Applied: https://github.com/astral-sh/uv/pull/12201/commits/e4a2fa1c1335af89e960036526d8456baef25fd9

jtfmumm commented on crates/uv/src/commands/project/run.rs:
```diff
@@ -1519,3 +1555,112 @@ fn read_recursion_depth_from_environment_variable() -> anyhow::Result<u32> {
+            "python4",
+            "python",
+            "python3.11.3",
+            "python39", // Still a valid executable, although likely a typo
```
Since this is likely a typo and we probably don't have to worry about the distant future at the moment, should we detect cases over 37 and treat them as typos? Not necessary for this PR though

JWLee89 replied:
I think this would be a good idea. 
Should we display a different kind of message if Python < 3.8 is used? 

Would it be okay if I create a separate PR for this once we discuss and decide on specifications here?

zanieb commented on crates/uv/src/commands/project/run.rs:
```diff
@@ -1519,3 +1555,112 @@ fn read_recursion_depth_from_environment_variable() -> anyhow::Result<u32> {
+/// - ❌ "python3abc", "python3.12b3", "", "python-foo"
+fn is_python_executable(executable_command: &str) -> bool {
+    executable_command
+        .strip_prefix("python")
```
Worth noting this isn't sufficient for Windows, where you'll have a `.exe` suffix.

JWLee89 replied:
Thank you for the followup!

Good point! I don't a windows machine available to me right now, but my guess is that the access pattern for invoking python will be something like the following: 

```Shell
# versioned executable
python3.11.exe
# General 
python.exe
```

Would this be correct? 
If so, can't we just strip the suffix `.exe`?

From some quick ChatGPT search, I found the following: 

|   Command   |                                Notes                               |
|:-----------:|:------------------------------------------------------------------:|
| python      | After installing via the official installer, adds to PATH.         |
| py          | Windows launcher for Python. Supports version flags like py -3.11. |
| python.exe  | Explicit executable, useful in scripts or batch files.             |
| py -X.Y     | e.g., py -3.10 to launch a specific version via Python launcher.   |
| pythonw.exe | Launches Python without a console window (GUI apps).               |

Out of these items above, do we need to support other forms such as `py`? Just wanted to be sure :)

JWLee89 replied:
Applied here: https://github.com/astral-sh/uv/pull/12201/commits/e6a801d0fe3deb62318b1e5aed5a18a3e350517b

Specs include: 
- strip `.exe` suffix
- Support absolute path python. E.g. `/Users/username/repos/uv/venv/bin/python3.10` 

Since spec was defined on the fly, might be good to review and discuss this together. 

Thank you! 🙇

JWLee89 committed:
refactor: rename version_part to specified_version

JWLee89 committed:
refactor: update code to be more idiomatic

JWLee89 committed:
feat: add support for .exe and absolute path python

jtfmumm committed:
Add snapshot tests for running missing Python versions

jtfmumm committed:
snapshot tests fixes

jtfmumm commented:
I pushed some snapshot (integration) tests for a handful of cases. 

I think we might want to update the error messages for `uv run pythonX.X.X` (the case you called out in your description). At least when it matches the patch version in use, it's a bit confusing. Though it's nice to have the rest of the message when it's not matched. But it seems like an opportunity to indicate that this is not expected usage. The first snapshot test I added captures this behavior.

jtfmumm commented:
> I pushed some snapshot (integration) tests for a handful of cases.

Looks like a couple of the snapshot tests I added are failing on macOS (but not Ubuntu or Windows) in CI. These pass for me locally. I'll have to investigate this further.

jtfmumm committed:
Ensure system Python isn't used in snapshot tests

JWLee89 commented:
> I pushed some snapshot (integration) tests for a handful of cases.

Thank you so much for adding the snapshot (integration) tests! 🙇 
I will keep this in mind when working on future PRs to add integration tests when necessary. 

> I think we might want to update the error messages for `uv run pythonX.X.X` (the case you called out in your description). At least when it matches the patch version in use, it's a bit confusing. Though it's nice to have the rest of the message when it's not matched. But it seems like an opportunity to indicate that this is not expected usage. The first snapshot test I added captures this behavior.

I absolutely agree and thank you for bringing this up. I should have clarified about this point properly in this PR. 
In my environment, I reproduced it with the following command: 

```bash
uv run python3.13.2
```

Results:

```bash
`python3.13.2` not available in the virtual environment, which uses python `3.13.2`. Did you mean to search for a Python 3.13.2 environment with `uv run -p 3.13.2 python`?
```

## Description of update

When users specify up to the patch version: 

```bash
uv run python3.3.3
```

I updated so that we get the following message.

```bash
error: Failed to spawn: `python3.3.3`
  Caused by: Please omit patch version. Try: `uv run python3.3`. Did you mean to search for a Python 3.3.3 environment with `uv run -p 3.3.3 python`?
```

However, if we omit patch version, we will get the following (same behavior as before): 

```bash
error: Failed to spawn: `python3.3`
  Caused by: `3.3` not available in the virtual environment, which uses python `3.13.2`. Did you mean to search for a Python 3.3 environment with `uv run -p 3.3 python`?
```

JWLee89 committed:
feat: provide meaningful error message when python patch version is provided

JWLee89 commented:
Hi @jtfmumm and @zanieb , 

Sorry for the ping ~ I know that you are both very busy. 
Just wanted to touch base and see if there are any changes / improvements that should be made to this PR 😄 . 

P.S. No rush, please respond only when you have time to review.

Thank you!

jtfmumm commented:
Hi @JWLee89. Thanks for your patience on this PR. We still need to make a decision on the messaging. I'll follow up with you.

JWLee89 commented:
> Hi @JWLee89. Thanks for your patience on this PR. We still need to make a decision on the messaging. I'll follow up with you.

Hi @jtfmumm ,

Thank you for the update! 👍
